netket/graph/lattice.py

```diff
@@ -362,24 +374,30 @@ def rotation_perm(self, period, axes=[0, 1]):
 
     def reflection_perm(self, axis=0):
         perm = []
-        for coord in self._atoms_coord:
-            new_coord = coord.copy()
-            new_coord[axis] = -1 * new_coord[axis]
+        ref_coords = self._coords.copy()
+        ref_coords[:,axis] = -1 * ref_coords[:,axis]
+        ref_coords = _np.ndarray.astype(1e5*_np.around(ref_coords,5),int)
+
+        ref_hash = {
+            hash(element.tobytes()): index
+            for index, element in enumerate(ref_coords)
+        }
 
+        for i, old_coord in enumerate(self._coords):
             searching = 1
             while searching:
-                for i, old_coord in enumerate(self.atoms_coord):
-                    shift_lattice = product(
-                        range(-1, 2), repeat=len(self._lattice_dims)
+                shift_lattice = product(
+                    range(-1, 2), repeat=len(self._lattice_dims)
+                )
+                for shift in shift_lattice:
+                    move_coord = old_coord + _np.sum(
+                        _np.expand_dims(_np.asarray(shift), 1) * self._lattice_dims,
+                        0,
```

Comment:
You could eliminate this bit by hashing the fractional coordinates of the sites in the simulation box. You can calculate that by multiplying the position vector with the inverse of the matrix made up of the vectors `basis_vectors[i]*extent[i]` and taking the fractional part. 
This guarantees you find matching sites even if the box is oddly shaped (now you can't know if something just maps outside the 4^d copies you generate) and gets rid of the 4^d-long loop.

Reply:
Also I tend to think reciprocal lattice vectors and the like will come useful when implementing #658

Reply:
I really like the hashing solution btw!

Reply:
Also, since there is a lot of code duplication with the hashing, you might want to have a method that takes a bunch of vectors and outputs their hashes. You might even want to cache the hashes of the original positions!

Reply:
Thanks, that’s an elegant solution

Reply:
I somehow missed this. Yeah I will fix that